### ISSUE TITLE
Identify Multiple Solutions for Deployment

### ISSUE THREAD

**Issue by pandevim** [YouTubeNLP/Backend#3]
Now that we've created a basic REST API and data schema design is still in progress (check #1), it's time to decided how we're going to deploy our servers and static builds. The purpose of this task is to enumerate potential solutions. That would primarly be:
1. Client Side Code: The static files/images created by the frontend framework (if any).
2. Server Side Code: The server (machine) which are going the power the NLP API and services.

Currently it's only going to be one platform. Deploying `development` and `production` builds separately would be focused later. 
## Deliverable
For each solution, "spike" the feasibility in the comment section. This may involve building a prototype, asking mentors, analsing scope of use case, steps to integrate with project, etc.

---
This would be owned by the @YouTubeNLP/backend team, although @YouTubeNLP/frontend can recommend or suggest anytime.

**Comment by edward2018211** [comment]
1. Client Side Code Solution: I propose to use AWS S3 to serve our static files/images since it is scalable and gives more control than say GitHub Pages. This is important mostly because our YouTube NLP site may need to hold tons of files as we are scraping from YouTube. More control is preferred since we aren't exactly sure what we may need to do in the future and giving ourselves more leeway would probably be a good choice. One of the downsides to using AWS S3 though is that we could incur costs, while other solutions like GitHub Pages is completely free. I don't think GitHub Pages lacks too much behind S3 for our use case, but if finances are a big factor, we might want to consider using GitHub Pages instead.

2. Server Side Code Solution: I believe that Google Firebase will be a good solution to run our server side code. It provides cloud functionality which fits us well since we really only need to power our endpoints and a few services. It's also has a cheaper cost than say EC2. While I think EC2 might fit our use case, I also think that since EC2 is a much more heavy setup with lots of functionality that I don't think we'll need to use, going with Firebase I think is the better option here. I believe for integrating with out project, we will simply need to do some authentication and create our cloud functions and we should be able to get it working.

**Comment by pandevim** [comment]
That's a insightful proposal. Although the examples I used above are only *examples*, we can go beyond and maybe end up using something else too. Definitely AWS S3 fits well in our case and later in the project we may need to use it. Google firebase has many advantages but the drawback I know are; since it is a serverless solution we won't get to use full resources of the machine.

My spike would be:
1. **Client**: Currently the frontend team has decided to use [Next.js](https://nextjs.org/) as their primary framework so using [Vercel](https://vercel.com/) would be a logical option since there both are created by the same team and work out of the box, also Vercel has a [Github Actions](https://github.com/marketplace/actions/vercel-action) present so that makes it very comfortable to setup.
2. **Server**: Since we need lot of computing resources (i.e. scraping, caching, ml prediction models services running in the background all the time), we ought to choose a good cloud infrastructure platform like [AWS EC2](https://aws.amazon.com/ec2/) but, it is paid and initially we don't need to invest that much. Other option can be [Heroku](https://www.heroku.com/) which has free plan.

**Comment by aviralch** [comment]
I was speaking with a mentor, and they suggested we might initially just have a local static folder (for Client Side). After we have an MVP we can then shift to a online service. This might help us write out rest of the code first and might save some money as well (not really sure about this).

**Comment by pandevim** [comment]
@aviralch yes that is correct, usually it is either `build`, `dist` or `doc` directory, but, since our project organization is different we ought to have some kind of hosting setup in the Frontend repo, that is we have different teams and repos like [NLP](https://github.com/YouTubeNLP/NLP) and [Backend](https://github.com/YouTubeNLP/Backend) so for a **Frontend** dev he/she can just spin up the local live server and can see how the website is looking but that is not possible perhaps inconvenient from a dev from other team who is working on other repo to monitor progress. Also I addressed some free solutions above in my [comment](https://github.com/YouTubeNLP/Backend/issues/3#issuecomment-652840646).

**Comment by edward2018211** [comment]
@pandevim I think both the client and server solutions you proposed seem to work pretty well. I agree that Vercel looks like a good choice since it is created by same people as Next.js. AWS EC2 also seems to work well since we can do a more with it compared to the Firebase I was talking about earlier.

**Comment by pandevim** [comment]
![ytnlp-architecture](https://user-images.githubusercontent.com/31156696/89115373-43c68f00-d4a5-11ea-9881-3ada47fd074d.png)
This is the current architecture of the project. Probably MongoDB will rule out as a solution, there are other dedicated hash table data libraries and data structure available like redis.
